src/realm/group_writer.cpp

```diff
@@ -301,25 +302,295 @@ GroupWriter::MapWindow* GroupWriter::get_window(ref_type start_ref, size_t size)
+    for (auto& entry : m_top_ref_map) {
+        if (unreachables.count(entry.first))
+            continue;
+        if (entry.first < m_oldest_reachable_version)
+            continue;
+        auto e =
+            std::make_unique<FreeList>(m_alloc, entry.second.top_ref, entry.second.logical_file_size, entry.first);
+        old_freelists.push_back(std::move(e));
+    }
+
+
+    // little helper: get a youngest version older than given
+    auto get_earlier = [&](uint64_t version) -> FreeList* {
+        auto pred = [](const std::unique_ptr<FreeList>& e, uint64_t v) {
+            return e->version < v;
+        };
+        auto it = std::lower_bound(old_freelists.begin(), old_freelists.end(), version, pred);
+        // There will allways be at least one freelist:
+        REALM_ASSERT(it != old_freelists.end());
+        if (it == old_freelists.begin())
+            return nullptr;
+        if ((*it)->version >= version)
+            --it;
+        REALM_ASSERT((*it)->version < version);
+        return it->get();
+    };
+
+
+    // find youngest time stamp in any block which (partially) overlaps a given one.
+    // if no such block exists, return the current version of the given block.
+    auto find_cover_for = [&](FreeSpaceEntry& entry, FreeList* it) -> uint64_t {
+        if (!it->resolved) {
+            // setup arrays
+            it->resolved = true;
+            if (it->top_ref) {
+                Array top_array(m_alloc);
+                top_array.init_from_ref(it->top_ref);
+                if (top_array.size() > Group::s_free_version_ndx) {
+                    // we have a freelist with versioning info
+                    it->present = true;
+                    it->positions.init_from_ref(top_array.get_as_ref(Group::s_free_pos_ndx));
+                    it->lengths.init_from_ref(top_array.get_as_ref(Group::s_free_size_ndx));
+                    it->versions.init_from_ref(top_array.get_as_ref(Group::s_free_version_ndx));
+                }
+            }
+        }
+        size_t index = 0, limit = 0;
+        if (it->present) {
+            // find block to allow for possible backdating
+            index = it->positions.upper_bound_int(entry.ref);
+            limit = it->positions.size();
+            if (index > 0) {
+                --index;
+            }
+        }
+        ref_type start_pos = 0, end_pos = 0;
+        uint64_t found_version = 0;
+        if (index < limit) {
+            start_pos = static_cast<ref_type>(it->positions.get(index));
+            end_pos = start_pos + static_cast<ref_type>(it->lengths.get(index));
+            if (end_pos <= entry.ref) {
+                index = limit; // discard entry
+            }
+        }
+        if (index < limit) {
+            // coallesce any subsequent contiguous entries
+            auto next = index + 1;
+            found_version = it->versions.get(index);
+            while (next < limit && it->positions.get(next) == (int64_t)end_pos) {
```

Comment:
Should you not stop when `end_pos` is bigger than the end of the entry?

Reply:
Good point. Though it'll rarely make a difference.